[SVG2] getPointAtLength should throw exception when in non-rendered document for SVGPathElement

Ahmad-S792 opened the pull request:
#### 9f2facf9ef6db9cf652f3e8729ce4e5cc3b90e26
<pre>
[SVG2] getPointAtLength should throw exception when in non-rendered document for SVGPathElement

<a href="https://bugs.webkit.org/show_bug.cgi?id=264876">https://bugs.webkit.org/show_bug.cgi?id=264876</a>
<a href="https://rdar.apple.com/problem/118720630">rdar://problem/118720630</a>

Reviewed by Said Abou-Hallawa.

This patch aligns WebKit with Gecko / Firefox, Blink / Chromium and Web-Specification [1]:

[1] <a href="https://svgwg.org/svg2-draft/types.html#InterfaceSVGGeometryElement">https://svgwg.org/svg2-draft/types.html#InterfaceSVGGeometryElement</a>

NOTE: SVGPathElement interface with SVGGeometryElement.

This patch aligns WebKit to throw exception in case of document being non-rendered element.

&quot;If current element is a non-rendered element, and the UA is not able to compute the total length
of the path, then throw an InvalidStateError.&quot;

I took the opportunity to add meaningful error to `SVGGeometryElement` as well and add check for
content-visibility to SVGPathElement.

* Source/WebCore/svg/SVGGeometryElement.cpp:
(SVGGeometryElement::getPointAtLength): Added meaningful message
* Source/WebCore/svg/SVGPathElement.cpp:
(SVGPathElement::getPointAtLength): Aligned with &apos;SVGGeometryElement&apos;
* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGGeometryElement.getPointAtLength-03-expected.txt: Rebaselined
* LayoutTests/svg/dom/SVGPolygonElement-baseVal-list-removal-crash.html: Rebaselined
* LayoutTests/svg/dom/path-pointAtLength.html: Ditto
* LayoutTests/svg/dom/path-pointAtLength-expected.txt: Ditto
* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGGeometryElement.getPointAtLength-04-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGGeometryElement.getPointAtLength-05-expected.txt: Rebaselined

Canonical link: <a href="https://commits.webkit.org/274308@main">https://commits.webkit.org/274308@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8200f0c82ae58abe633f8d863f02f7a0f2f2cd99

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38544 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17475 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40869 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41079 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34221 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40849 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20266 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14818 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32406 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39117 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14725 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33532 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12792 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12776 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34388 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42355 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35011 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34810 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38603 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13384 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11064 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36809 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14984 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/33689 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8672 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13851 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14458 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/svg/SVGPathElement.cpp:
```diff
@@ -183,6 +183,11 @@ ExceptionOr<Ref<SVGPoint>> SVGPathElement::getPointAtLength(float distance) cons
 
+    auto* renderer = this->renderer();
+    // Spec: If current element is a non-rendered element, throw an InvalidStateError.
+    if (!renderer)
```
No need for the local variable since it is only used in this if-statement:
```
// Spec: If current element is a non-rendered element, throw an InvalidStateError.
if (!renderer())
```

Ahmad-S792 replied:
I will update it for `SVGGeometryElement` as well. :-)

Ahmad-S792 replied:
Actually - in `SVGGeometryElement`, it is used multiple times, so I will leave it as is.

Ahmad-S792 replied:
Done in [237f550](https://github.com/WebKit/WebKit/commit/237f550c0ea8d94136d85245cda340cbd0bd720d).

Ahmad-S792 committed:
[SVG2] getPointAtLength should throw exception when in non-rendered document for SVGPathElement

https://bugs.webkit.org/show_bug.cgi?id=264876
rdar://problem/118720630

Reviewed by Said Abou-Hallawa.

This patch aligns WebKit with Gecko / Firefox, Blink / Chromium and Web-Specification [1]:

[1] https://svgwg.org/svg2-draft/types.html#InterfaceSVGGeometryElement

NOTE: SVGPathElement interface with SVGGeometryElement.

This patch aligns WebKit to throw exception in case of document being non-rendered element.

"If current element is a non-rendered element, and the UA is not able to compute the total length
of the path, then throw an InvalidStateError."

I took the opportunity to add meaningful error to `SVGGeometryElement` as well and add check for
content-visibility to SVGPathElement.

* Source/WebCore/svg/SVGGeometryElement.cpp:
(SVGGeometryElement::getPointAtLength): Added meaningful message
* Source/WebCore/svg/SVGPathElement.cpp:
(SVGPathElement::getPointAtLength): Aligned with 'SVGGeometryElement'
* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGGeometryElement.getPointAtLength-03-expected.txt: Rebaselined
* LayoutTests/svg/dom/SVGPolygonElement-baseVal-list-removal-crash.html: Rebaselined
* LayoutTests/svg/dom/path-pointAtLength.html: Ditto
* LayoutTests/svg/dom/path-pointAtLength-expected.txt: Ditto
* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGGeometryElement.getPointAtLength-04-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGGeometryElement.getPointAtLength-05-expected.txt: Rebaselined

Canonical link: https://commits.webkit.org/274308@main
